### ISSUE TITLE
Support Write Concern Results

### ISSUE THREAD

**Issue by nscavell** [vert-x3/vertx-mongo-client#2]
Starting with version 2.6, Mongo returns write concern results of write operations. The Mongo driver returns things like UpdateResult and InsertResult. We should return something similar.

**Comment by karianna** [comment]
Closed #40 as a duplicate.  John's comments from there:

A number of the operations such as update and replace throw away useful info by taking the UpdateResult (or whatever), throwing it away then returning null. Such as at:

https://github.com/vert-x3/vertx-mongo-client/blob/master/vertx-mongo-client/src/main/java/io/vertx/ext/mongo/impl/MongoClientImpl.java#L153

**Comment by cescoffier** [comment]
I'm not sure this will be fix in the 3.1.0, probably 3.2.0. What do you think @karianna.

**Comment by karianna** [comment]
Agreed, shifting.

**Comment by cescoffier** [comment]
Shifting again.

**Comment by MazizEsa** [comment]
Any one is working on this?

**Comment by karianna** [comment]
@MazizEsa Nope, please go for it!

**Comment by MazizEsa** [comment]
Will give ticket a try.

**Comment by MazizEsa** [comment]
@karianna 

> A number of the operations such as update and replace throw away useful info by taking the UpdateResult (or whatever), throwing it away then returning null. Such as at:

https://github.com/vert-x3/vertx-mongo-client/blob/master/vertx-mongo-client/src/main/java/io/vertx/ext/mongo/impl/MongoClientImpl.java#L153

I notice inserts return id field (https://github.com/vert-x3/vertx-mongo-client/blob/master/vertx-mongo-client/src/main/java/io/vertx/ext/mongo/impl/MongoClientImpl.java#L126). But according to https://docs.mongodb.org/manual/reference/method/db.collection.insert/ it will return either WriteResult or BulkWriteResult.

Im just wondering what's the use case for the id returned from insert call.

**Comment by karianna** [comment]
I assume people use it for a follow up query or as 'proof' that the insert succeeded

**Comment by MazizEsa** [comment]
If that's the case we can just return WriteResults(BulkWriteResult) instead. Which brings me the next question as to why mongo-driver (https://github.com/mongodb/mongo-java-driver/blob/master/driver-async/src/main/com/mongodb/async/client/MongoCollectionImpl.java#L318) return null on its callback. Any idea?

**Comment by karianna** [comment]
Not sure on that one sorry - I'd also check with @johnoliver as well

Cheers,
Martijn

**Comment by MazizEsa** [comment]
Will put this on hold. Will inquire the vertx group.

**Comment by jyemin** [comment]
The callback type for the insert methods is SingleResultCallback<Void> so the only thing they can return is null.  The type is Void because there is no useful information to return in reply to a successful insert.

That said, it may prove to be a mistake an have been better to return an InsertResult (with no properties) as a placeholder for any future metadata that may be returned.  You may want to protect yourself in that way.

**Comment by MazizEsa** [comment]
@jyemin. Sorry for my vague question. What I wanted to ask was why the callback type is SingleResultCallback<Void>? From the documents (https://docs.mongodb.org/manual/reference/method/db.collection.insert/#db.collection.insert), insert would return either WriteResult or BulkResult which contains the write concern stuffs. Seeing other operations(Update, replace and delete) it will return SingleResultCallback type accordingly (UpdateResult, or deleteResult).

**Comment by jyemin** [comment]
@MazizEsa :  If you look at UpdateResult and BulkWriteResult, none of their properties really apply to an insert in a way that would be useful for insertOne or insertMany.  If the insert method returns normally, the number if documents inserted is 1 for insertOne and the size of the document list for insertMany.  If the insert method returns abnormally by throwing an exception, the exception will indicate how many documents were actually inserted (see MongoBulkWriteException).

**Comment by MazizEsa** [comment]
@jyemin thanks. Understood.

**Comment by MazizEsa** [comment]
Still working on this thing.

@karianna Im planning to go fourth and make another method in the MongoClient interface with (Update, replace and etc.) with AsyncResult type to respective operation results (UpdateResult, DeleteResult etc.) Am I missing anything that may render this direction pointless or is there any other way that wouldn't have to make any changes on the interface?

**Comment by karianna** [comment]
@MazizEsa - that seems like a valid approach.  @johnoliver - any comments?

**Comment by MazizEsa** [comment]
Proceed with the implementation.

**Comment by MazizEsa** [comment]
@karianna since Iill be modifying the interface that's also be used for other language shims generation, there are some limitation as to the data type that can be used with the AsyncResult (https://github.com/vert-x3/vertx-codegen#permitted-types).

Since the affected methods are update, replace and remove, I would propose the following:

1. Update and replace will return update result, therefore, in our case we will return JsonObject containing matched, upserted and modified as per https://docs.mongodb.org/manual/reference/method/db.collection.update/#writeresults-update

1.  Remove will return delete result, therefor we will return JsonObject containing removed as per https://docs.mongodb.org/manual/reference/method/db.collection.remove/#writeresults-remove.

Whenever both encounter error or write concern error it will return in the exception as MongoBulkWriteException.

Any inputs?

**Comment by johnoliver** [comment]
I am not sure why you need to add any methods to MongoClient. Did you mean MongoCollection? If so can you say what methods you want to add?

I have not looked closely at this recently so may not appreciate all the corner cases here, but in general it seems to me that we should simply convert the WriteResult/BulkWriteResult to some JsonObject or better yet a proper domain class representation that is a DataObject and return that. 

So methods like:
```
public void insertOne(final TDocument document, final SingleResultCallback<Void> callback) 
```

would change to 

```
public void insertOne(final TDocument document, final SingleResultCallback<MongoWriteResult> callback) 
```

Where we can come up with an appropriate MongoWriteResult that allows us to return everything we need such as generated ids, update counts etc, (it will probably look very similar to mongos WriteResult/BulkWriteResult data).

This is going to be an API breaking change which is not great, but its a step in the right direction.

Regarding nothing on the WriteResult being useful, I disagree, what about doing an upsert? The write result will tell you if you have created a new record or modified one.

**Comment by jyemin** [comment]
It's not that nothing on the WriteResult is useful, it's that (currently) nothing on the WriteResult applies to an insert (an upsert is considered a type of update, not an insert).

**Comment by johnoliver** [comment]
Ahh, yes you are right, specifically for insert, there is nothing that cannot be inferred from the other data. I would still probably return some useful data if we are going to do it for the others though just to be consistent

**Comment by MazizEsa** [comment]
@johnoliver Im thinking something like for instance update and updateWithOptions, I will add another updateWithResult and updateWithOptionsWithResult and deprecated the older two.

**Comment by MazizEsa** [comment]
@jyemin quick one. When you call get("_id") (https://github.com/mongodb/mongo-java-driver/blob/master/driver-core/src/main/com/mongodb/connection/WriteCommandResultHelper.java#L98) do you get BSONArray or a single BsonNumber or it could be anything?

**Comment by jyemin** [comment]
It can be an instance of any BsonValue subclass.

**Comment by MazizEsa** [comment]
@jyemin don't you have any specified contract as what mongo will return in this case for upsert id? I could go on handling all the cases but it's a tad overboard.

**Comment by jyemin** [comment]
If the MongoDB server generates the _id, then it will always be an ObjectId.  But more typically, the user will be choosing the _id value in the client application, and it may or may not be an ObjectId.  Consider the following snippet:

`        
    UpdateResult updateResult = mongoCollection.updateOne(
                Filters.eq("_id", 1),
                Updates.inc("x", 1),
                new UpdateOptions().upsert(true));
`

In this case the upsertedId will be of type BsonInt32.

**Comment by MazizEsa** [comment]
@jyemin i see. So can I use this https://docs.mongodb.org/manual/indexes/#indexes as a guide?

**Comment by jyemin** [comment]
@MazizEsa Can you elaborate?  I'm not clear how that link is relevant to our discussion.

**Comment by MazizEsa** [comment]
@jyemin I was wondering if that document can be used as a reference point to determine the possible values (range) that can be set to the _id.

**Comment by jyemin** [comment]
_Any_ BSON value (even including null) can be used for the _id of a document.

**Comment by MazizEsa** [comment]
@jyemin understood. Appreciate your input.

**Comment by joseaio** [comment]
Please, anyone can confirm if this feature must be in vertx 3.3.0 version?

**Comment by karianna** [comment]
@joseaio We're not sure yet sorry.

**Comment by karianna** [comment]
Closed via #78 - thanks @MazizEsa
